source/common/json/json_loader.cc

```diff
@@ -2,254 +2,475 @@
-    }
-    return member_itr->value.GetDouble();
+int64_t Factory::Field::getInteger(const std::string& name, int64_t default_value) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr != value_.object_value_.end()) {
+    return getInteger(name);
+  } else {
+    return default_value;
   }
+}
 
-  double getDouble(const std::string& name, double default_value) const override {
-    if (!value_.HasMember(name.c_str())) {
-      return default_value;
+ObjectPtr Factory::Field::getObject(const std::string& name, bool allow_empty) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr == value_.object_value_.end()) {
+    if (allow_empty) {
+      return createObject();
     } else {
-      return getDouble(name);
+      throw Exception(fmt::format("key '{}' missing", name));
     }
+  } else if (!value_itr->second->isType(Type::Object)) {
+    throw Exception(fmt::format("key '{}' not an object", name));
+  } else {
+    return value_itr->second;
   }
+}
 
-  uint64_t hash() const override {
-    rapidjson::StringBuffer buffer;
-    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
-    value_.Accept(writer);
-    return std::hash<std::string>{}(buffer.GetString());
+std::vector<ObjectPtr> Factory::Field::getObjectArray(const std::string& name) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr == value_.object_value_.end() || !value_itr->second->isType(Type::Array)) {
+    throw Exception(fmt::format("key '{}' missing or not an array", name));
   }
 
-  void iterate(const ObjectCallback& callback) const override {
-    for (auto& member : value_.GetObject()) {
-      ObjectImplBase object(member.value, member.name.GetString());
-      bool need_continue = callback(member.name.GetString(), object);
-      if (!need_continue) {
-        break;
-      }
-    }
+  std::vector<FieldPtr> array_value = value_itr->second->arrayValue();
+  return {array_value.begin(), array_value.end()};
+}
+
+std::string Factory::Field::getString(const std::string& name) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr == value_.object_value_.end() || !value_itr->second->isType(Type::String)) {
+    throw Exception(fmt::format("key '{}' missing or not a string", name));
   }
+  return value_itr->second->stringValue();
+}
 
-  bool hasObject(const std::string& name) const override { return value_.HasMember(name.c_str()); }
+std::string Factory::Field::getString(const std::string& name,
+                                      const std::string& default_value) const {
+  checkType(Type::Object);
+  auto value_itr = value_.object_value_.find(name);
+  if (value_itr != value_.object_value_.end()) {
+    return getString(name);
+  } else {
+    return default_value;
+  }
+}
```

Comment:
These getters are all kind of boiler platish, consider a template function implementation.